gk/fib.c

```
@@ -1664,3 +1666,385 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+}
+
+static int
+fillup_gk_fib_dump_entry(struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+#define CTYPE_STRUCT_GK_FIB_DUMP_ENTRY_PTR "struct gk_fib_dump_entry *"
+
+static int
+list_ipv4_fib_entries(lua_State *l, struct gk_lpm *ltbl,
+	uint32_t ctypeid_dump_entry)
+{
+	int ret, index;
+
+	void *cdata;
+
+	struct gk_fib *fib;
+	const struct rte_lpm_rule *re4;
+	struct rte_lpm_iterator_state state;
+	struct gk_fib_dump_entry dentry;
+
+	ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		lua_pushstring(l , "gk: failed to initialize the lpm rule iterator state");
+		goto out;
+	}
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv4;
+		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
+		dentry.len = state.depth;
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0)
+			goto out;
+
+		cdata = luaL_pushcdata(l, ctypeid_dump_entry,
+			sizeof(struct gk_fib_dump_entry *));
+		*(struct gk_fib_dump_entry **)cdata = &dentry;
+
+		lua_pushvalue(l, 1);
+		lua_insert(l, 1);
+
+		if (lua_pcall(l, 1, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		ret = lua_tonumber(l, -1);
+		if (ret < 0)
+			goto out;
```

Let the Lua handler just raise an error instead of having to return an error number. This comment also applies to `list_ipv6_fib_entries()`.